### PR TITLE
feat(popup-menu): add API to provide custom no search results entry

### PR DESCRIPTION
**Pull request by philippfromme**
Adds `getEmptyPlaceholder` API to `PopupMenuProvider` to customize what's shown when there are no entries (after search). Based on [the original issue](https://github.com/camunda/camunda-bpmn-js/issues/322) is clear that a simple string or even an HTML string as it's used for icons (`imgHtml`) won't cut it here which means we have to allow users to return a `VNode`. The search is passed to the callback returning the `VNode` so it can be taken into account.

### Example

brave_CVRyjayZfB

----

Closes https://github.com/bpmn-io/diagram-js/issues/846

### CODE REVIEW

File: test/spec/features/popup-menu/PopupMenuSpec.js [hunk 1]
```diff
@@ -2191,8 +2216,8 @@ function getGroup(groupName) {
   return domQuery('[data-group="' + groupName + '"]', getPopupContainer());
 }
 
-function whenStable() {
-  return new Promise(resolve => setTimeout(resolve, 300));
+function sleep(ms = 300) {
```
Comment (nikku) [inline test/spec/features/popup-menu/PopupMenuSpec.js:2219]: I'm not sure this rename is sensible. At the end we want to wait until things (preact for that matter) is updated (or stable to inspect).

Comment (philippfromme) [reply test/spec/features/popup-menu/PopupMenuSpec.js:2219]: I found `whenStable` really confusing. I will revert that change.

File: lib/features/popup-menu/PopupMenuComponent.js [hunk 2]
```diff
@@ -246,7 +249,7 @@ export default function PopupMenuComponent(props) {
           />
         </div>
         ${ entries.length === 0 && html`
-          <div class="djs-popup-no-results">No matching entries found.</div>
+          <div class="djs-popup-no-results">${ noSearchResultsEntry || 'No matching entries found.' }</div>
```
Comment (nikku) [inline lib/features/popup-menu/PopupMenuComponent.js:252]: Idea: Provide the default in `PopupEditor` so we can translate the string, too.

Comment (philippfromme) [reply lib/features/popup-menu/PopupMenuComponent.js:252]: Can you explain? 🤔

Comment (nikku) [reply lib/features/popup-menu/PopupMenuComponent.js:252]: Other idea: Consider if we want to compute the result dynamically:

```javascript
<div class="djs-popup-no-results">${ noResultsEntry(value) }</div>
```

Comment (nikku) [reply lib/features/popup-menu/PopupMenuComponent.js:252]: Consider I want to add an awesome box:

```html
We did not find an entry matching `<foo>`, do you want to <a href="https://some-market-palce-api?search=${foo}`">Market place</a>
```

Comment (philippfromme) [reply lib/features/popup-menu/PopupMenuComponent.js:252]: I was wondering if something like that would be useful. Will try that out.

**Comment by philippfromme** [comment]
@nikku I adjusted the implementation as suggested. The search is now passed to the callback, which returns the custom node.

**Comment by nikku** [comment]
:upside_down_face: How looks did out? Did you complete fine tuning this PR?

**Comment by philippfromme** [comment]
@nikku Should be ready to be reviewed now. What a pain in the 🍑 that was.

**Comment by nikku** [comment]
@philippfromme As discussed I went forth and do a little bit more work on this PR:

* Now we need to expose an `getEmptyPlaceholder` hook to be used for integration. 
* The first hook (by priority) wins and will be taken.
* We don't ship a default empty placeholder, but instead rely on integrators to ship it. This solves the issue that all but this string were already translatable, and translation usually happens in integrating libraries. We can discuss this as I found one additional, not translated string: `Open element documentation`.

Please checkout what I did and see if it makes sense.

File: lib/features/popup-menu/PopupMenuComponent.js [hunk 3]
```diff
@@ -81,18 +84,17 @@ export default function PopupMenuComponent(props) {
...
-        .join('---')
-        .toLowerCase();
+      ].map(string => string.toLowerCase());
 
+      // every word of `value` should be included in one of the searchable fields
       return value
         .toLowerCase()
         .split(/\s/g)
-        .every(term => search.includes(term));
+        .every(word => searchableFields.some(field => field.includes(word)));
```
Comment (nikku) [inline lib/features/popup-menu/PopupMenuComponent.js:97]: Could make sense to test cover this.

Comment (philippfromme) [reply lib/features/popup-menu/PopupMenuComponent.js:97]: Previously, we matched non-existent strings:

brave_BG1L16Jg0d

This commit fixed it: https://github.com/bpmn-io/diagram-js/pull/847/commits/7c67215376ba2cb1e1430e8a67aca9358df9c5a4

I added tests as well: https://github.com/bpmn-io/diagram-js/pull/847/commits/baa739194463edd919203fb6dfeb0987ebae0aab

**Comment by philippfromme** [comment]
Looks good to me. @nikku I requested another review so feel free to approve and merge.

File: lib/features/popup-menu/PopupMenuComponent.js (hunk 3 above)
Comment (nikku) [reply lib/features/popup-menu/PopupMenuComponent.js:97]: Thanks. Fine with me, though the intention was to build the most simplest version possible (and assume no user would type `---` in the search box) :see_no_evil:.

**Review by nikku — APPROVED** [review]

File: lib/features/popup-menu/PopupMenuComponent.js (hunk 3 above)
Comment (philippfromme) [reply lib/features/popup-menu/PopupMenuComponent.js:97]: Well, I did. 🤡

**Comment by nikku** [comment]
We are not here to "explain linking", but to identify errors (and layout the basic rules).

Documentation tries to pickup the user, i.e. clearly tell what wrong usage exists: https://github.com/bpmn-io/bpmnlint/pull/128/files#diff-1c0cd3d147700fbecbfc2f14e57a1438fd5f920c628975153a9b6e9ab29629b0.

**Pull request merged**